execution_profile: whitelist/blacklist filtering

muzarski opened the pull request:
Fixes: https://github.com/scylladb/cpp-rust-driver/issues/243

This PR introduces whitelist/blacklist host filtering to the driver.

## Filtering rules
There are currently 4 filtering rules (we could introduce 2 more in the future - rack filtering):
- ip whitelist
- ip blacklist
- dc whitelist
- dc blacklist

The logic for checking if a host is accepted is following (in the specified order):
1. if `ip whitelist` is non-empty, and if it does not contain host's ip -> reject
2. if `ip blacklist` is non-empty, and if it contains host's ip -> reject
3. if `dc whitelist` is non-empty, and if it does not contain host's dc -> reject
4. if `dc blacklist` is non-empty, and if it contains host's dc -> reject

## HostFilter
Since all of the execution profiles are defined before session creation (in other words: user cannot define new profiles after session was created and connected), cpp-driver (and so, cpp-rust-driver as well) can see which hosts are disabled by all execution profiles.

If there is a host which is rejected by **all** execution profiles (including the default one), the driver does not open the connection to such host. This is exactly what we do in this PR - after collecting all of the filtering rules of each profile, we compute the filtering rules for our custom `HostFilter` implementor - namely `CassHostFilter`. This is done by computing the unions of whitelists and intersection of blacklists.

## Vec or HashSet
For now, I decided to use `Vec` to represent the set of items in the whitelists/blacklists. I'd say that we do not expect them to contain a lot of items, so I think `Vec` is more efficient than `HashSet` in this specific case. One could argue that this code is not critical (and I agree) - this is a configuration phase. I'm open to suggestions.

## Integration tests

### ExecutionProfile suite
I have not enabled any yet - they require `cass_future_coordinator`, which is used to check the ip of host the request was routed to. TBH, I think this is a blocker for merging - I prefer to wait for the `cass_future_coordinator`, and enable the corresponding tests (and maybe implement some additional tests on my own). I still believe that this PR is ready for review, though.

### HostFilter suite
This is the suite I implemented myself - currently there are two tests that reject all nodes using execution profiles - in result, session does not have any connections to route the requests to.

### DisconnectedNullStringApiArgs suite
Enabled remaining 4 test cases from this suite - they check whether driver behaves as expected when null string is provided to filtering config methods.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have implemented Rust unit tests for the features/changes introduced.
- [ ] I have enabled appropriate tests in `.github/workflows/build.yml` in `gtest_filter`.
- [ ] I have enabled appropriate tests in `.github/workflows/cassandra.yml` in `gtest_filter`.

Lorak-mmk commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -13,6 +18,55 @@ pub(crate) struct FilteringConfig {
+        })
+    });
+
+    if let Some(item_iter) = item_iter {
+        // We should append new entries.
+        filtering_vec.extend(item_iter);
+    } else {
+        // If the string is empty or null, we should clear the list.
+        filtering_vec.clear();
+    }
+
```
Is that really how cpp-driver does it?

muzarski replied:
https://github.com/scylladb/cpp-driver/blob/master/src/cluster_config.cpp#L350-L361

(Notice that `SAFE_STRLEN` returns 0 for null pointer)

muzarski replied:
Since `cass_cluster_set_whitelist_filtering(cluster, nullptr)` clears the list (because of `SAFE_STRLEN(nullptr) == 0`), I decided that `cass_cluster_set_whitelist_filtering_n(cluster, nullptr, <non_zero_length>)` should clear the list as well (as opposed, to cpp-driver, which tries to create a `String` from null pointer).

Lorak-mmk commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -144,3 +219,60 @@ pub(crate) enum LoadBalancingKind {
+            .on_request_failure(request, latency, node, error);
+    }
+
+    fn name(&self) -> String {
+        "FilteringLoadBalancingPolicy".to_string()
+    }
```
We could include a name of child policy here too.

Lorak-mmk commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -291,3 +291,110 @@ impl HostFilter for CassHostFilter {
 }
+
+/// Returns the union of all vectors (sets).
+/// If the union is empty, it returns None.
+fn nonempty_union<'a, T>(iter: impl Iterator<Item = &'a Vec<T>>) -> Option<Vec<T>>
+where
+    T: Clone + Ord + 'a,
+{
+    let mut union = Vec::new();
+
+    for values in iter {
+        for v in values {
+            if !union.contains(v) {
+                union.push(v.clone());
+            }
+        }
+    }
+
+    (!union.is_empty()).then_some(union)
+}
+
+/// Returns the intersection of all vectors (sets).
+/// If the intersection is empty, it returns None.
+fn nonempty_intersection<'a, T>(mut iter: impl Iterator<Item = &'a Vec<T>>) -> Option<Vec<T>>
+where
+    T: Clone + PartialEq + 'a,
+{
+    // Get the first set (initial intersection).
+    let mut intersection = iter.next().cloned();
+
+    // Remove the items from the intersection that are not present in the other sets.
+    if let Some(intersection) = intersection.as_mut() {
+        for set in iter {
+            intersection.retain(|item| set.contains(item));
+        }
+    }
+
+    // If the intersection is empty, return None.
+    intersection.filter(|v| !v.is_empty())
+}
```
Commit: lbp: introduce utility methods for set operations 

I think that with host filtering the performance of Vec vs HashSet is not relevant (this is not even close to hot path). What we should look at is readability. Wouldn't methods present in HashSet / BTreeSet help with implementation?

muzarski replied:
On a high level, there is no `HashSet` API that would allow us to implement these as a "one-liners". The logic for intersection would remain the same. However, the logic for union could be simplified - we can convert each set to iterator, then flatten and collect to `HashSet`. Currently, we need a loop, so we don't insert the duplicates.

I have nothing against transitioning to `HashSet` - as you said - this is not close to hot path.

muzarski replied:
cc: @wprzytula

Lorak-mmk replied:
Wouldn't some combination of `fold` (or similar Iterator methods) and `union` / `intersection`  HashSet methods allow to express both of those more cleanly?

muzarski replied:
`union` and `intersection` methods take two `HashSet`s and return an iterator. I don't see a neat way to generalize it and reuse these methods to implement `union`/`intersection` for multiple sets (without needless `collect()`ing in `fold`). I think that the `retain` method is clean and efficient at the same time.

muzarski commented:
I have already found a bug in my `HostFilter` filtering rules computation. Consider following scenario:
We have two nodes: "ip1" and "ip2". Let's say we have two execution profiles, one with the whitelist being "ip1", the other with empty (disabled) whitelist. In result, the `HostFilter` whitelist will be a union of these -> "ip1". In result, we filter out the "ip2" node - even though it is **accepted** by at least one execution profile (the one with empty whitelist).

I'll fix it - the `nonempty_union` should return `None`, if at least one of the provided sets is empty (i.e. one of the whitelist rules is disabled). This proves that we need the tests before merging this.

Note: the `nonempty_intersection` looks good in such scenario, because the if at least one of the blacklists is empty, the resulting `HostFilter` blacklist will be empty (disabled) as well.

muzarski commented:
Rebased on master.

muzarski commented:
v2:
- addressed @Lorak-mmk comments (haven't transitioned to `HashSet` yet - still to be discussed)
- fixed the bug I mentioned above
- introduced `HostFilterTest` suite and implemented two test cases where we disable all nodes (driver has no connections to route the queries).

muzarski commented:
v2.1: Previously I implemented new tests, but forgot to enable them..

Lorak-mmk reviewed (approved):
Looks good to me. I don't really care if we use Vec or HashSet, so I'm approving now. If you decide to change the container, then I'll re-approve.

muzarski commented:
Rebased on master

wprzytula commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -237,3 +237,120 @@ impl HostFilter for CassHostFilter {
     }
 }
+
+/// Returns the union of all non-empty vectors (sets).
+/// If at least one set is empty, it return None.
+fn nonempty_union<'a, T>(iter: impl Iterator<Item = &'a Vec<T>>) -> Option<Vec<T>>
+where
+    T: Clone + Ord + 'a,
+{
```
:question: Why do we require `Ord`? Isn't `Eq` enough?

wprzytula commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -282,6 +282,57 @@ where
+    /// Computes the filtering rules from the load balancing policies, and
+    /// returns a corresponding `HostFilter`.
+    ///
+    /// In cpp-driver, the following rule is upheld: if a host is rejected
+    /// by **all** policies, the connection to that host is not opened at all.
```
:wrench: Please provide a link to the relevant piece of code in cpp-driver.

wprzytula commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -282,6 +282,57 @@ where
+                .map(|lbp_config| &lbp_config.filtering.whitelist_dc),
+        );
+
+        let blacklist_dc = nonempty_intersection(
+            configs
+                .clone()
+                .map(|lbp_config| &lbp_config.filtering.blacklist_dc),
+        );
```
:pick: Here we don't have to `clone()` - it's the last use of `configs`.

wprzytula commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -124,6 +128,7 @@ impl CassSessionInner {
         }
 
         let session = session_builder
+            .host_filter(host_filter)
             .build()
```
:thought_balloon: Maybe, as an optimisation, only provide a `HostFilter` to the `Session` if there is any nonempty whitelist/blacklist?

wprzytula replied:
I suppose minority of users will be using the whitelist/blacklist functionality, and even fewer of them will disallow some hosts in all execution profiles and in the cluster.

Lorak-mmk replied:
> 💭 Maybe, as an optimisation, only provide a HostFilter to the Session if there is any nonempty whitelist/blacklist?

Why? HostFilter is only called in places where such micro-optimizations won't matter.

wprzytula replied:
Because why not? Why even turn on a filter that passes through everything unconditionally?

Lorak-mmk replied:
> Because why not?

Because it requires additional logic.

muzarski replied:
I'm resolving this - not planning to add this micro-optimization

muzarski commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -13,6 +16,73 @@ pub(crate) struct FilteringConfig {
+impl FilteringInfo {
+    /// Checks if the host is valid according to the filtering rules.
+    ///
+    /// If host does not belong to any datacenter, it's datacenter is treated
+    /// as empty string. This way, if for example only `dc1` is whitelisted, the
+    /// node with unknown DC will be rejected.
```
> ❓
> 
>     1. Can an empty datacenter be specified in the whitelist or in the blacklist?

No - `explode` function in cpp-driver filters out empty strings.

>     2. If not, is this semantics really correct? Consider a node that provides no DC info. Then we cannot blacklist it, so the only viable solution to disallow it is to whitelist all other DCs.

You are correct - we cannot blacklist "no info" dc. But this is the semantics from cpp-driver - I just sticked to it. We could always change it if necessary.

muzarski commented on scylla-rust-wrapper/src/cluster.rs:
```diff
@@ -1754,6 +1754,132 @@ mod tests {
+                        .is_empty()
+                );
+            }
+
+            // Populate the list again...
+            {
+                cass_cluster_set_blacklist_filtering(
+                    cluster_raw.borrow_mut(),
+                    c"1.1.1.1,2.2.2.2,foo,,,,  ,3.3.3.3,".as_ptr(),
+                );
+
+                let cluster = BoxFFI::as_ref(cluster_raw.borrow()).unwrap();
+                assert_eq!(
+                    cluster.load_balancing_config.filtering.blacklist_hosts,
+                    vec![
+                        IpAddr::V4(Ipv4Addr::new(1, 1, 1, 1)),
+                        IpAddr::V4(Ipv4Addr::new(2, 2, 2, 2)),
+                        IpAddr::V4(Ipv4Addr::new(3, 3, 3, 3))
+                    ]
+                );
+            }
+
+            // ..and clear it with the null pointer
+            {
+                cass_cluster_set_blacklist_filtering(cluster_raw.borrow_mut(), std::ptr::null());
+
+                let cluster = BoxFFI::as_ref(cluster_raw.borrow()).unwrap();
+                assert!(
+                    cluster
+                        .load_balancing_config
+                        .filtering
+                        .blacklist_hosts
+                        .is_empty()
+                );
+            }
+
+            cass_cluster_free(cluster_raw);
+        }
+    }
+
```
True, I was lazy. I decided that testing cluster methods is enough, since both cluster and exec_profile methods are just wrappers over `update_comma_delimited_list`, which contains most of the tested logic.

Should I add these?

wprzytula replied:
Let's add them. They should mostly involve copying and adjusting existing tests for cluster methods.

muzarski commented on scylla-rust-wrapper/src/cluster.rs:
```diff
@@ -1089,6 +1089,142 @@ pub unsafe extern "C" fn cass_cluster_set_latency_aware_routing_settings(
+            hosts_size,
+            |s| match IpAddr::from_str(s) {
+                Ok(ip) => Some(ip),
+                Err(err) => {
+                    tracing::warn!("Failed to parse ip address <{}>: {}", s, err);
+                    None
+                }
```
The remaining "good" entries are appended anyway.

wprzytula replied:
`error` does not necessarily mean a complete failure. It may signify that just part of the operation failed. IMO, passing an unparsable IP address counts as an error.

muzarski replied:
Ok

muzarski commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -13,6 +16,73 @@ pub(crate) struct FilteringConfig {
+    /// If host does not belong to any datacenter, it's datacenter is treated
+    /// as empty string. This way, if for example only `dc1` is whitelisted, the
+    /// node with unknown DC will be rejected.
+    pub(crate) fn is_host_valid(&self, ip: &IpAddr, dc: &Option<String>) -> bool {
```
Since we use `&str` in case `None` is provided (thus, we can't use `contains()` Vec method anyway, because it requires `&String`), I agree.

wprzytula replied:
Good catch about `Vec::contains()` - I didn't consider that.

muzarski commented on scylla-rust-wrapper/src/load_balancing.rs:
```diff
@@ -90,3 +165,60 @@ pub(crate) enum LoadBalancingKind {
     },
 }
+
+#[derive(Debug)]
+pub(crate) struct FilteringLoadBalancingPolicy {
+    pub(crate) filtering: FilteringInfo,
+    pub(crate) child_policy: Arc<dyn LoadBalancingPolicy>,
+}
+
+impl LoadBalancingPolicy for FilteringLoadBalancingPolicy {
+    fn pick<'a>(
+        &'a self,
+        request: &'a RoutingInfo,
+        cluster: &'a ClusterState,
+    ) -> Option<(scylla::cluster::NodeRef<'a>, Option<scylla::routing::Shard>)> {
+        let picked = self.child_policy.pick(request, cluster);
+
+        picked.and_then(|target| {
+            let node = target.0;
+            self.filtering
+                .is_host_valid(&node.address.ip(), &node.datacenter)
+                .then_some(target)
+        })
+    }
+
+    fn fallback<'a>(
+        &'a self,
+        request: &'a RoutingInfo,
+        cluster: &'a ClusterState,
+    ) -> FallbackPlan<'a> {
+        Box::new(
+            self.child_policy
+                .fallback(request, cluster)
+                .filter(|(node, _shard)| {
+                    self.filtering
+                        .is_host_valid(&node.address.ip(), &node.datacenter)
+                }),
+        )
+    }
```
> Black- and whitelisting DCs must be done in a layer above

Correct - unless we introduce the support for it in rust-driver (some configuration on `DefaultPolicy`) - this is not that important IMO.

> will likely often make `pick()` return `None` and thus force `fallback()` to execute

Do you think this will happen often? Why would someone prefer some dc, and then filter out the node from preferred dc.

But maybe you meant the general scenario, where underlying LBP is non-dc-aware. In such case - that's true, `pick` may often return unwanted node.

But I think the general use case for filtering is using a dc-aware LBP, and (on top of it) filter out some nodes from remote dcs which we don't want to fallback to.

wprzytula replied:
> > will likely often make `pick()` return `None` and thus force `fallback()` to execute
> 
> Do you think this will happen often? Why would someone prefer some dc, and then filter out the node from preferred dc.
> 
> But maybe you meant the general scenario, where underlying LBP is non-dc-aware. In such case - that's true, `pick` may often return unwanted node.

Yes, I meant the non-DC-aware underlying LBP.
> 
> But I think the general use case for filtering is using a dc-aware LBP, and (on top of it) filter out some nodes from remote dcs which we don't want to fallback to.

Such use case makes sense. However, another one seems to me just as as this one: there are 2 local DCs and 3 remote DCs. We want to filter out the remote DCs. OTOH, we don't want to prefer any local DC over the other local DC. Then we are left with this problem, because `DefaultPolicy` will often `pick()` targets from remote DCs, so `fallback()` will be inevitable.

wprzytula commented:
:question: You haven't marked:
```
I have enabled appropriate tests in .github/workflows/build.yml in gtest_filter.
I have enabled appropriate tests in .github/workflows/cassandra.yml in gtest_filter.
```
Were them just overlooked?

muzarski commented:
> ❓ You haven't marked:
> 
> ```
> I have enabled appropriate tests in .github/workflows/build.yml in gtest_filter.
> I have enabled appropriate tests in .github/workflows/cassandra.yml in gtest_filter.
> ```
> 
> Did you just overlook them?

The `ExecutionProfile` tests that are testing filtering methods require support for exposing/enforcing coordinator. I'm waiting for these to be implemented. Then, we can enable the corresponding tests and merge this RP (I don't want to merge it without tests enabled).

muzarski commented:
v2.2: Addressed @wprzytula comments (some of them still need an answer)

muzarski commented:
v2.3: Next iteration of addressing @wprzytula comments. As a bonus (after discussion on Slack), I added casts to `fn(&str) -> T` for closures with empty env.

muzarski commented:
Thanks for the approves - I'm not merging yet. I'll wait for `cass_future_coordinator` support to enable the corresponding integration tests.

Lorak-mmk commented:
So probably it is a tradeoff of inlining vs less copies, right? In other words, with casts we get 2 less copies, but the functions won't be inlined. Do we have reasons to believe this is the right tradeoff to do here?

muzarski commented:
> Do we have reasons to believe this is the right tradeoff to do here?

I don't have any. Maybe @wprzytula has some.

wprzytula commented:
> So probably it is a tradeoff of inlining vs less copies, right? In other words, with casts we get 2 less copies, but the functions won't be inlined. Do we have reasons to believe this is the right tradeoff to do here?

By telling the compiler that this is the same function, we leave the decision up to it whether to inline or not.
By not saying it, the compiler may still inline (and then the code ends up equivalent) or not. If not, this is the worst case: there are multiple copies of the same function, each called from another place in code. With our hint, all calls can target the same function.

muzarski commented:
v3: Rebased on master (with getting coordinator support) and enabled 4 integration tests from `[DC]ExecutionProfileTest` suites. (See only last 3 commits)
